### PR TITLE
Add blockchain event notifier service

### PR DESCRIPTION
**Pull request by enzoc4**
# Add a `Notifier` service

## relates to #2354 

- [X] Add websocket endpoint
- [X] Process block events and send message
- [X] Process tip change event and send message
- [x] Add automated test (technically done, but still needs some though)
- [X] Add setting to limit the number of connections (and set a default)
- [X] Define the message format, right now they are ~~just strings manually crafted (`New Block: hash` `New Tip: hash`)~~ json strings with serde. Maybe we can provide binary too depending on the client
- [ ] Add this to documentation.
- [X] Introduce some idea of versioning the API? Something like
    - [X] Add a version in the endpoint path, maybe the same that the rest api.
    - [ ] Use websocket subprotocols
- [ ] Send connection close event when shutting down

The max_connection setting is under the `Rest` settings, for example:

```yaml
rest:
  listen: "127.0.0.1:8443"
  notifier:
      max_connections: 5
```
### Questions

`NewBlock` events are always sent even when `NewTip`  has the same hash. I'm not sure if we want to filter these or not.

It can be manually quick-tested by opening a websocket from a browser, for example:

```html
<!DOCTYPE html>
<html>
    <head>
        <title>Notifier reader</title>
    </head>
    <body>
        <h1>new blocks</h1>
        <div id="status">
            <p><em>Connecting...</em></p>
        </div>
        <script type="text/javascript">
        var uri = 'ws://localhost:8443/notifier';
        var ws = new WebSocket(uri);
        function message(data) {
            var line = document.createElement('p');
            line.innerText = data;
            var status = document.getElementById('status');
            status.appendChild(line);
        }
        ws.onopen = function() {
            var status = document.getElementById('status');
            status.innerHTML = "<p><em>Connected!</em></p>";
        }
        ws.onmessage = function(msg) {
            message(msg.data);
        };
        </script>
    </body>
</html>

```

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -2110,6 +2160,39 @@ dependencies = [
  "serde_yaml",
 ]
 
+[[package]]
+name = "openssl"
```
Comment (NicolasDP) [inline Cargo.lock:2164]: we will have problems on some architecture with this

File: jormungandr/src/rest/mod.rs [hunk 2]
```diff
@@ -32,11 +33,13 @@ pub async fn start_rest_server(config: Rest, explorer_enabled: bool, context: Co
...
+    let notifier = notifier::filter(context.clone());
     if explorer_enabled {
         let explorer = explorer::filter(context);
-        setup_cors(api.or(explorer), config, stopper_rx).await;
+        setup_cors(api.or(notifier).or(explorer), config, stopper_rx).await;
     } else {
-        setup_cors(api, config, stopper_rx).await;
+        setup_cors(api.or(notifier), config, stopper_rx).await;
     }
 }
```
Comment (NicolasDP) [inline jormungandr/src/rest/mod.rs:44]: the API is going to be off. the notifier is part of the `v1` API.

File: testing/jormungandr-integration-tests/Cargo.toml [hunk 3]
```diff
@@ -40,6 +40,7 @@ url = "2.1.1"
 yaml-rust = "0.4.4"
 indicatif = "0.15.0"
 fs_extra = "1.1.0"
+tungstenite = "0.11.0"
```
Comment (NicolasDP) [inline testing/jormungandr-integration-tests/Cargo.toml:43]: hmm okay, so this is what brings in all the dependencies

File: jormungandr/src/notifier/mod.rs [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    max_connections: usize,
+}
+
+#[derive(Serialize)]
+enum JsonMessage {
+    #[serde(serialize_with = "to_hex")]
+    NewBlock(HeaderId),
+    #[serde(serialize_with = "to_hex")]
+    NewTip(HeaderId),
+}
```
Comment (NicolasDP) [inline jormungandr/src/notifier/mod.rs:32]: This would live in better in jormungandr-lib

File: jormungandr/src/notifier/mod.rs [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+async fn process_message(
+    logger: Logger,
+    clients: Arc<tokio::sync::Mutex<Clients>>,
+    msg: Message,
+    mut disconnected: MessageBox<usize>,
+) {
+    let warp_msg = JsonMessage::from(msg).into();
+    let dead = async move { notify_all(clients, warp_msg).await };
```
Comment (NicolasDP) [inline jormungandr/src/notifier/mod.rs:111]: what is wrong with simply using:

```suggestion
    let dead = notify_all(clients, warp_msg).await;
```

File: jormungandr/src/notifier/mod.rs [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+// error codes in 4000-4999 are reserved for private use.
+// I couldn't find an error code for max connections, so I'll use the first one for now
+// maybe using the standard error code for Again is the right thing to do
+const MAX_CONNECTIONS_ERROR_CLOSE_CODE: u16 = 4000;
+const MAX_CONNECTIONS_ERROR_REASON: &str = "MAX CONNECTIONS reached";
+
+#[derive(Clone)]
+pub struct Notifier {
+    next_user_id: Arc<AtomicUsize>,
+    clients: Arc<tokio::sync::Mutex<Clients>>,
```
Comment (NicolasDP) [inline jormungandr/src/notifier/mod.rs:22]: hmm... not going to scale well.

File: Cargo.lock (hunk 1 above)
Comment (enzoc4) [reply Cargo.lock:2164]: good point, tungstenite is only used in the tests, so I think turning off the tls feature may be fine.

File: jormungandr/src/notifier/mod.rs (hunk 5 above)
Comment (enzoc4) [reply jormungandr/src/notifier/mod.rs:111]: nothing, just a leftover of a little refactor

File: jormungandr/src/notifier/mod.rs (hunk 4 above)
Comment (enzoc4) [reply jormungandr/src/notifier/mod.rs]: moved

File: jormungandr/src/rest/mod.rs (hunk 2 above)
Comment (enzoc4) [reply jormungandr/src/rest/mod.rs]: actually, it was under `/notifier` and not `/v0/notifier`. But I moved it to `v1/notifier`.

**Comment by eugene-babichenko** [comment]
The overall idea is good but I'd go for a different approach to send messages to clients. Namely, having an actor that collects all events from other actors, builds WS messages right away and just broadcast these messages to client connection handlers. As an optimization you could skip serialization when there is no listeners.

Such approach would allow to perform the message serialization only once per each message. Now it is done `n` times where `n` is the number of subscriptions. This will also simplify the WS handling code. In fact, this code will become universal, because all handlers will receive the same type of messages.

**Comment by enzoc4** [comment]
@eugene-babichenko I'm not sure I get your point. The difference I see between the current code and what you are describing is the 

> builds WS messages right away

part. Which I think I get, we can move the serialization before/up in the flow, right now we are just serializing ids, so it's probably not that expensive, but makes sense.

Maybe you are referring to something different, though.

**Comment by eugene-babichenko** [comment]
@enzoc4 

I mean the following data flow:

```
+----------------+                                                       +-----------+
| block events   +--------------+                      +---------------->+ ws 1      |
+----------------+      +-------v-------+      +---------------+         +-----------+
                        | make message  +----->+ serialize msg |
+----------------+      +-------^-------+      +---------------+         +-----------+
| tip   events   +--------------+                      +---------------->+ ws 2      |
+----------------+                                                       +-----------+
```

**Comment by enzoc4** [comment]
> @enzoc4
> 
> I mean the following data flow:
> 
> ```
> +----------------+                                                       +-----------+
> | block events   +--------------+                      +---------------->+ ws 1      |
> +----------------+      +-------v-------+      +---------------+         +-----------+
>                         | make message  +----->+ serialize msg |
> +----------------+      +-------^-------+      +---------------+         +-----------+
> | tip   events   +--------------+                      +---------------->+ ws 2      |
> +----------------+                                                       +-----------+
> ```

I think now the data flow is just like that, but I don't see how does it simplify any code tbh. The websocket handlers were always receiving text anyway. It still makes sense to serialize at once, probably.

**Pull request closed without merging**